File: backend-django/events/notification_utils.py
from django.utils import timezone

from emails import send_event_capacity_alert
from .models import EventStatus


NOTIFICATION_SNAPSHOT_FIELDS = [
    'date_start',
    'date_end',
    'address_full',
    'address_city',
    'address_country',
    'address_visibility',
    'address_reveal_date',
    'online_platform',
    'online_link',
    'online_visibility',
    'online_reveal_date',
]


def capture_event_notification_snapshot(event):
    return {field: getattr(event, field) for field in NOTIFICATION_SNAPSHOT_FIELDS}


def get_event_update_messages(previous, event):
    changes = []

    if previous['date_start'] != event.date_start or previous['date_end'] != event.date_end:
        changes.append(
            "La date ou l'horaire a été mis à jour "
            f"(nouveau créneau : {event.date_start.strftime('%d/%m/%Y à %H:%M')} "
            f"→ {event.date_end.strftime('%d/%m/%Y à %H:%M')})."
        )

    address_fields = ['address_full', 'address_city', 'address_country']
    if any(previous[field] != getattr(event, field) for field in address_fields):
        changes.append("Le lieu de l'événement a été mis à jour.")

    online_fields = ['online_platform', 'online_link']
    if any(previous[field] != getattr(event, field) for field in online_fields):
        changes.append("Les informations de connexion en ligne ont été mises à jour.")

    return changes


def reset_scheduled_notification_flags(event, previous):
    update_fields = []

    if previous['date_start'] != event.date_start:
        for field in [
            'reminder_7d_sent_at',
            'reminder_1d_sent_at',
            'reminder_3h_sent_at',
            'organizer_digest_sent_at',
        ]:
            if getattr(event, field) is not None:
                setattr(event, field, None)
                update_fields.append(field)

    address_related_fields = [
        'address_full',
        'address_city',
        'address_country',
        'address_visibility',
        'address_reveal_date',
    ]
    if any(previous[field] != getattr(event, field) for field in address_related_fields):
        if event.address_reveal_email_sent_at is not None:
            event.address_reveal_email_sent_at = None
            update_fields.append('address_reveal_email_sent_at')

    online_related_fields = [
        'online_platform',
        'online_link',
        'online_visibility',
        'online_reveal_date',
    ]
    if any(previous[field] != getattr(event, field) for field in online_related_fields):
        if event.online_reveal_email_sent_at is not None:
            event.online_reveal_email_sent_at = None
            update_fields.append('online_reveal_email_sent_at')

    return update_fields


def notify_event_capacity_milestones(event):
    if event.status != EventStatus.PUBLISHED or event.unlimited_capacity or not event.capacity:
        update_fields = []
        for field in ['almost_full_notified_at', 'full_notified_at']:
            if getattr(event, field) is not None:
                setattr(event, field, None)
                update_fields.append(field)
        if update_fields:
            event.save(update_fields=update_fields)
        return

    from registrations.models import RegistrationStatus

    confirmed_count = event.registrations.filter(status=RegistrationStatus.CONFIRMED).count()
    fill_ratio = confirmed_count / event.capacity
    now = timezone.now()
    update_fields = []

    if confirmed_count < event.capacity and event.full_notified_at is not None:
        event.full_notified_at = None
        update_fields.append('full_notified_at')
    if fill_ratio < 0.8 and event.almost_full_notified_at is not None:
        event.almost_full_notified_at = None
        update_fields.append('almost_full_notified_at')

    if confirmed_count >= event.capacity:
        if event.full_notified_at is None:
            send_event_capacity_alert(event, 'FULL')
            event.full_notified_at = now
            update_fields.append('full_notified_at')
    elif fill_ratio >= 0.8 and event.almost_full_notified_at is None:
        send_event_capacity_alert(event, 'ALMOST_FULL')
        event.almost_full_notified_at = now
        update_fields.append('almost_full_notified_at')

    if update_fields:
        event.save(update_fields=update_fields)

File: backend-node/src/utils/notificationUtils.js
'use strict';

/**
 * notificationUtils.js — Utilitaires de notifications planifiées pour les événements.
 *
 * Équivalent Node.js de events/notification_utils.py (Django).
 */

const { sendEventCapacityAlert } = require('../services/emailService');

const NOTIFICATION_SNAPSHOT_FIELDS = [
  'date_start',
  'date_end',
  'address_full',
  'address_city',
  'address_country',
  'address_visibility',
  'address_reveal_date',
  'online_platform',
  'online_link',
  'online_visibility',
  'online_reveal_date',
];

/**
 * Capture les champs importants d'un event avant une modification.
 * @param {object} event — instance Sequelize Event
 * @returns {object}
 */
function captureEventNotificationSnapshot(event) {
  const snapshot = {};
  for (const field of NOTIFICATION_SNAPSHOT_FIELDS) {
    snapshot[field] = event[field];
  }
  return snapshot;
}

/**
 * Compare un snapshot pris avant modification avec l'event modifié.
 * Retourne une liste de messages de changement (en français).
 * @param {object} previous — snapshot avant modification
 * @param {object} event — instance Sequelize Event après modification
 * @returns {string[]}
 */
function getEventUpdateMessages(previous, event) {
  const changes = [];

  const fmtDate = (d) => d
    ? new Date(d).toLocaleDateString('fr-FR', {
        day: '2-digit', month: '2-digit', year: 'numeric',
        hour: '2-digit', minute: '2-digit',
      })
    : '—';

  const prevStart = previous.date_start ? String(new Date(previous.date_start)) : null;
  const prevEnd = previous.date_end ? String(new Date(previous.date_end)) : null;
  const newStart = event.date_start ? String(new Date(event.date_start)) : null;
  const newEnd = event.date_end ? String(new Date(event.date_end)) : null;

  if (prevStart !== newStart || prevEnd !== newEnd) {
    changes.push(
      `La date ou l'horaire a été mis à jour (nouveau créneau : ${fmtDate(event.date_start)} → ${fmtDate(event.date_end)}).`
    );
  }

  const addressFields = ['address_full', 'address_city', 'address_country'];
  if (addressFields.some(f => previous[f] !== event[f])) {
    changes.push("Le lieu de l'événement a été mis à jour.");
  }

  const onlineFields = ['online_platform', 'online_link'];
  if (onlineFields.some(f => previous[f] !== event[f])) {
    changes.push("Les informations de connexion en ligne ont été mises à jour.");
  }

  return changes;
}

/**
 * Réinitialise les flags de notifications planifiées si les champs concernés ont changé.
 * Appelle event.update() si des champs doivent être réinitialisés.
 * @param {object} event — instance Sequelize Event (déjà sauvegardé)
 * @param {object} previous — snapshot avant modification
 */
async function resetScheduledNotificationFlags(event, previous) {
  const updates = {};

  const prevStart = previous.date_start ? String(new Date(previous.date_start)) : null;
  const newStart = event.date_start ? String(new Date(event.date_start)) : null;

  if (prevStart !== newStart) {
    for (const field of ['reminder_7d_sent_at', 'reminder_1d_sent_at', 'reminder_3h_sent_at', 'organizer_digest_sent_at']) {
      if (event[field] !== null) updates[field] = null;
    }
  }

  const addressRelated = ['address_full', 'address_city', 'address_country', 'address_visibility', 'address_reveal_date'];
  if (addressRelated.some(f => previous[f] !== event[f])) {
    if (event.address_reveal_email_sent_at !== null) updates.address_reveal_email_sent_at = null;
  }

  const onlineRelated = ['online_platform', 'online_link', 'online_visibility', 'online_reveal_date'];
  if (onlineRelated.some(f => previous[f] !== event[f])) {
    if (event.online_reveal_email_sent_at !== null) updates.online_reveal_email_sent_at = null;
  }

  if (Object.keys(updates).length > 0) {
    await event.update(updates);
  }
}

/**
 * Vérifie si l'event atteint un seuil de capacité (80% ou 100%) et envoie une alerte.
 * Gère aussi la réinitialisation des flags si la capacité redescend.
 * @param {object} event — instance Sequelize Event avec company chargée
 */
async function notifyEventCapacityMilestones(event) {
  if (event.status !== 'PUBLISHED' || event.unlimited_capacity || !event.capacity) {
    const updates = {};
    if (event.almost_full_notified_at !== null) updates.almost_full_notified_at = null;
    if (event.full_notified_at !== null) updates.full_notified_at = null;
    if (Object.keys(updates).length > 0) await event.update(updates);
    return;
  }

  const { Registration } = require('../models');
  const confirmedCount = await Registration.count({
    where: { event_id: event.id, status: 'CONFIRMED' },
  });
  const fillRatio = confirmedCount / event.capacity;
  const now = new Date();
  const updates = {};

  if (confirmedCount < event.capacity && event.full_notified_at !== null) {
    updates.full_notified_at = null;
  }
  if (fillRatio < 0.8 && event.almost_full_notified_at !== null) {
    updates.almost_full_notified_at = null;
  }

  if (confirmedCount >= event.capacity) {
    if (event.full_notified_at === null) {
      await sendEventCapacityAlert(event, 'FULL');
      updates.full_notified_at = now;
    }
  } else if (fillRatio >= 0.8 && event.almost_full_notified_at === null) {
    await sendEventCapacityAlert(event, 'ALMOST_FULL');
    updates.almost_full_notified_at = now;
  }

  if (Object.keys(updates).length > 0) await event.update(updates);
}

module.exports = {
  captureEventNotificationSnapshot,
  getEventUpdateMessages,
  resetScheduledNotificationFlags,
  notifyEventCapacityMilestones,
};
